onnx/defs/nn/defs.cc

```diff
@@ -2832,4 +2833,392 @@ ONNX_OPERATOR_SET_SCHEMA(
               return true;
             }));
+
+static const char* ScalarDotProductAttention_ver23_doc = R"DOC(
+
+Computes scaled dot product attention on query, key and value tensors, using an optional attention mask if passed.
+
+This operator covers self and cross variants of the attention operation based on sequence lengths of K, Q and V.
+For self attention, kv_sequence_length equals to q_sequence_length.
+For cross attention, query and key might have different lengths.
+
+This operator also covers the 3 following variants based on the number of heads:
+1) Multi-headed Attention (MHA): Described in the paper https://arxiv.org/pdf/1706.03762, q_num_heads = kv_num_heads.
+2) Group-query Attention (GQA): Described in the paper https://arxiv.org/pdf/2305.13245, q_num_heads > kv_num_heads.
+3) Multi-query Attention (MQA): Described in the paper https://arxiv.org/pdf/1911.02150, q_num_heads > kv_num_heads, q_num_heads=1.
+
+Attention bias to be added is calculated based on attn_mask input and is_causal attribute, only one of which can be provided.
+1) If is_causal is set to 1, the attention masking is a lower triangular matrix when the mask is a square matrix. The attention masking has the form of the upper left causal bias due to the alignment.
+2) attn_mask: A boolean mask where a value of True indicates that the element should take part in attention or a float mask of the same type as query, key, value that is added to the attention score.
+
+Both past and present state key/values are optional. They shall be used together, and not allowed to use only one of them.
+The following pattern is applied to the Q, K and V inputs after appropriate reshaping of K and V inputs based on sequence lengths and num heads provided:
+
+          Q          K          V
+          |          |          |
+          |      Transpose      |
+          |          |          |
+          ---MatMul---          |
+                |               |
+       scale---Mul              |
+                |               |
+     at_bias---Add              |
+                |               |
+             Softmax            |
+                |               |
+                -----MatMul------
+                        |
+                        Y
+
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    ScalarDotProductAttention,
+    23,
+    OpSchema()
+        .SetDoc(ScalarDotProductAttention_ver23_doc)
+        .Attr(
+            "is_causal",
+            "If set to 1, the attention masking is a lower triangular matrix when the mask is a square matrix. "
+            "The attention masking has the form of the upper left causal bias due to the alignment.",
+            AttributeProto::INT,
+            static_cast<int64_t>(0))
+        .Attr(
+            "scale",
+            "Scaling factor applied prior to softmax. Default value is 1/sqrt(head_size)",
+            AttributeProto::FLOAT,
+            OPTIONAL_VALUE)
+        .Attr(
+            "q_num_heads",
```

Comment:
Is the head_size same for q and kv? If so, would it be simpler to specify just the head_size, instead of num_heads for q and kv? Just a thought. We should think through the implications for (exporters, the function-definition, shape-inference, etc., since they could be impacted.)

Reply:
Is this for group query attention?